File: src/main/java/io/github/intellij/dlanguage/features/DChooseByNameContributor.java
package io.github.intellij.dlanguage.features;

import com.intellij.navigation.ChooseByNameContributor;
import com.intellij.navigation.NavigationItem;
import com.intellij.openapi.project.Project;
import com.intellij.psi.search.GlobalSearchScope;
import com.intellij.psi.stubs.StubIndex;
import com.intellij.util.ArrayUtil;
import io.github.intellij.dlanguage.psi.interfaces.DNamedElement;
import io.github.intellij.dlanguage.stubs.index.DAllNameIndex;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

/**
 * The "go to symbol"
 */
public class DChooseByNameContributor implements ChooseByNameContributor {
    @NotNull
    @Override
    public String[] getNames(final Project project, final boolean includeNonProjectItems) {
        return ArrayUtil.toStringArray(StubIndex.getInstance().getAllKeys(DAllNameIndex.KEY, project));
    }

    @NotNull
    @Override
    public NavigationItem[] getItemsByName(final String name, final String pattern,
                                           final Project project, final boolean includeNonProjectItems) {
        @NotNull final GlobalSearchScope scope = includeNonProjectItems ? GlobalSearchScope.allScope(project) : GlobalSearchScope.projectScope(project);
        final Collection<DNamedElement> classes = StubIndex.getElements(DAllNameIndex.KEY, name, project, scope, DNamedElement.class);
        final List<NavigationItem> items = new ArrayList<>(classes.size());
        items.addAll(classes);
        return items.toArray(new NavigationItem[items.size()]);
    }
}

File: src/main/kotlin/io/github/intellij/dlanguage/features/DChooseByNameContributor.kt
package io.github.intellij.dlanguage.features

import com.intellij.navigation.ChooseByNameContributor
import com.intellij.navigation.NavigationItem
import com.intellij.openapi.project.Project
import com.intellij.psi.search.GlobalSearchScope
import com.intellij.psi.stubs.StubIndex
import com.intellij.util.ArrayUtil
import io.github.intellij.dlanguage.psi.interfaces.DNamedElement
import io.github.intellij.dlanguage.stubs.index.DAllNameIndex

/**
 * The "go to symbol"
 */
class DChooseByNameContributor : ChooseByNameContributor {
    override fun getNames(project: Project, includeNonProjectItems: Boolean): Array<String> {
        return ArrayUtil.toStringArray(StubIndex.getInstance().getAllKeys(DAllNameIndex.KEY, project))
    }

    override fun getItemsByName(
        name: String, pattern: String,
        project: Project, includeNonProjectItems: Boolean
    ): Array<NavigationItem> {
        val scope =
            if (includeNonProjectItems) GlobalSearchScope.allScope(project) else GlobalSearchScope.projectScope(project)
        val classes = StubIndex.getElements(DAllNameIndex.KEY, name, project, scope, DNamedElement::class.java)
        val items: MutableList<NavigationItem> = ArrayList(classes.size)
        items.addAll(classes)
        return items.toTypedArray()
    }
}
